subsys/bluetooth/mesh/transport.c

```diff
@@ -1044,18 +1045,22 @@ static int trans_unseg(struct net_buf_simple *buf, struct bt_mesh_net_rx *rx,
 	hdr = net_buf_simple_pull_u8(buf);
 
 	if (rx->ctl) {
-		return ctl_recv(rx, hdr, buf, seq_auth);
-	}
-
-	if (buf->len < 1 + APP_MIC_LEN(0)) {
+		err = ctl_recv(rx, hdr, buf, seq_auth);
+	} else if (buf->len < 1 + APP_MIC_LEN(0)) {
 		LOG_ERR("Too short SDU + MIC");
-		return -EINVAL;
+		err = -EINVAL;
+	} else {
+		/* Adjust the length to not contain the MIC at the end */
+		buf->len -= APP_MIC_LEN(0);
+		err = sdu_recv(rx, hdr, 0, buf, &sdu, NULL);
 	}
 
-	/* Adjust the length to not contain the MIC at the end */
-	buf->len -= APP_MIC_LEN(0);
+	/* Update rpl only if there is place and upper logic accepted incoming data. */
+	if ((err == 0 || IS_ENABLED(CONFIG_BT_MESH_ACCESS_LAYER_MSG)) && rpl != NULL) {
```

Comment:
With this I disagree, I think it will be more clear if error code will be set to 0 when the access layer msg callback is called. It is bad to move access stuff to transport layer.

Reply:
Setting error code to 0 will not work in if unicast address is used.

1. There is no model opcode and access layer returns correct access layer opcode.
2. application expects this opcode over access layer callback and handles it.
3. transport sees returned value isn't zero and skips rpl update
4. replay attack on application is possible

I guess if feature `CONFIG_BT_MESH_ACCESS_LAYER_MSG` is enabled, then rpl should behave more differently.

Reply:
This has one more issue. If callback is not set, the message is not processed.

Reply:
Also if unconditionally to return 0 for broadcast messages then just spamming over broadcast any garbage will wipe out flash.

Reply:
> Setting error code to 0 will not work if unicast address is used.
> 
> 1. There is no model opcode and access layer returns correct access layer error.
> 2. application expects this opcode over access layer callback and handles it.
> 3. transport sees returned value isn't zero and skips rpl update
> 4. replay attack on application is possible
> 
> I guess if feature `CONFIG_BT_MESH_ACCESS_LAYER_MSG` is enabled, then rpl should behave more differently.

The error code must be overwritten in this case. This is why I said is should be set to 0 when the callback is called.

> Also if unconditionally to return 0 for broadcast messages then just spamming over broadcast any garbage will wipe out flash.

I'm not following. Please elaborate with concrete example. If a message is processed, it shall be stored in RPL.

Reply:
> > Setting error code to 0 will not work if unicast address is used.
> > 
> > 1. There is no model opcode and access layer returns correct access layer error.
> > 2. application expects this opcode over access layer callback and handles it.
> > 3. transport sees returned value isn't zero and skips rpl update
> > 4. replay attack on application is possible
> > 
> > I guess if feature `CONFIG_BT_MESH_ACCESS_LAYER_MSG` is enabled, then rpl should behave more differently.
> 
> The error code must be overwritten in this case. This is why I said is should be set to 0 when the callback is called.
> 

The error code rewriting for unicast messages will break oppcode aggregator model. Here: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/op_agg_srv.c#L84 
and here: 
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/op_agg_cli.c#L87

> > Also if unconditionally to return 0 for broadcast messages then just spamming over broadcast any garbage will wipe out flash.
> 
> I'm not following. Please elaborate with concrete example. If a message is processed, it shall be stored in RPL.

Sure. Scenario: 
There are two types of devices in the network. Devices A1 have model M with opcodes o1-o3. Devices A2 have model M with opcodes o1-o3 and application handler for o4.

Devices A2 support EMDS from NCS and do not care about flash wire leveling and send o4 over multicast or broadcast addresses.
Devices A1 receive o4 (do not do anything in both access and application) but update rpl every received o4.

Finally, communication between A2 devices wipes out physically devices A1. The service company has to send workers to change devices deployed in "the fields". Customer reports bug to us about broken flash despite devices were addresses very rarely.

Reply:
> The error code rewriting for unicast messages will break oppcode aggregator model. Here: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/op_agg_srv.c#L84
> and here:
> https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/op_agg_cli.c#L87

Good catch. Then it means the special check for opcode aggregator should be added as the encapsulated messages don't have own seqnum, only the opagg sequence and status messages.

> Devices A1 receive o4 (do not do anything in both access and application) but update rpl every received o4.

In your example I disagree with this statement in quote. If the callback is implemented, then the application does something as the message is provided to the application through the callback.

Reply:
> This has one more issue. If callback is not set, the message is not processed.

And you should also consider this comment. If callback is not set, but the option is enabled, we will always store the seqnum in RPL which we are trying to avoid with this PR.

Reply:
> > The error code rewriting for unicast messages will break oppcode aggregator model. Here: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/op_agg_srv.c#L84
> > and here:
> > https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/op_agg_cli.c#L87
> 
> Good catch. Then it means the special check for opcode aggregator should be added as the encapsulated messages don't have own seqnum, only the opagg sequence and status messages.

Didn't get what you mean. Please clarify.

> 
> > Devices A1 receive o4 (do not do anything in both access and application) but update rpl every received o4.
> 
> In your example I disagree with this statement in quote. If the callback is implemented, then the application does something as the message is provided to the application through the callback.

There are possible two cases:
1. Application doesn't provide any callback. This is possible. We do not mandate it.
2. Devices A1 have the disabled `CONFIG_BT_MESH_ACCESS_LAYER_MSG` at all.

If Access just returns error zero when rx broadcast o4 in both cases rpl will be updated. The worst situation here is that, both A1 and A2 have been developed correctly for its own purposes (probably even different companies). However, their coexistence in the single network is deadly for A1.

Reply:
> > This has one more issue. If callback is not set, the message is not processed.
> 
> And you should also consider this comment. If callback is not set, but the option is enabled, we will always store the seqnum in RPL which we are trying to avoid with this PR.

Fair enough. We should mention it in KConfig option. Customers should know about potential problems.

PS. Sorry for updating your comment. Didn't know github allows doing that. :)

Reply:
> Didn't get what you mean. Please clarify.

When calling the access layer msg callback we need to check if an opagg model processes encapsulated messages now. If so, don't overwrite the error value.

> 1. Application doesn't provide any callback. This is possible. We do not mandate it.
> 2. > Devices A1 have the disabled CONFIG_BT_MESH_ACCESS_LAYER_MSG at all.

In the first case messages are not processed, then it is OK to not store them.
The second point is not a problem we are discussing in this thread. It is clear what to do in this case.

> Fair enough. We should mention it in KConfig option. Customers should know about potential problems.

It is not good to provide tons of documentation for issues that could be solved in the code.

IMHO 1:
If a user enabled the access layer msg callback option AND set the callback, the seqnums should be stored in RPL even if their processing would result in error defined by spec section 3.7.3.4. Because if we don't do this, then thus customers who rely only on the callback will see these messages in the callback over and over again.

IMHO 2:
Introducing a dependency on the upper layer internal details (which is the access layer) in the lower layer (which is the transport layer) is not correct from the architecture perspective. The access layer shall return correct value which the transport layer should use to decide if the message was processed and it should be stored in RPL.

Reply:
Regarding the opagg and the callback, we can take this code https://github.com/zephyrproject-rtos/zephyr/blob/1e2b889bbb6929fe34ea6a81a05fcf29bb8a792b/subsys/bluetooth/mesh/access.c#L1536-L1560 out of the `bt_mesh_model_recv`, move it into another function and called it from the opagg models. The transport layer still calls `bt_mesh_model_recv`, but not the opagg models.
But then, when an opagg message is received, the access layer msg callback will be called only for the opagg message, and will not be called for the messages encapsulated messages.

Like this:
```
diff --git a/subsys/bluetooth/mesh/access.c b/subsys/bluetooth/mesh/access.c
index 328ec0dd281..8a818056892 100644
--- a/subsys/bluetooth/mesh/access.c
+++ b/subsys/bluetooth/mesh/access.c
@@ -1518,21 +1518,12 @@ static int element_model_recv(struct bt_mesh_msg_ctx *ctx, struct net_buf_simple
 	return ACCESS_STATUS_SUCCESS;
 }
 
-int bt_mesh_model_recv(struct bt_mesh_msg_ctx *ctx, struct net_buf_simple *buf)
+int bt_mesh_model_msg_recv(struct bt_mesh_msg_ctx *ctx, struct net_buf_simple *buf)
 {
 	int err = ACCESS_STATUS_SUCCESS;
 	uint32_t opcode;
 	uint16_t index;
 
-	LOG_DBG("app_idx 0x%04x src 0x%04x dst 0x%04x", ctx->app_idx, ctx->addr,
-		ctx->recv_dst);
-	LOG_DBG("len %u: %s", buf->len, bt_hex(buf->data, buf->len));
-
-	if (IS_ENABLED(CONFIG_BT_TESTING)) {
-		bt_test_mesh_model_recv(ctx->addr, ctx->recv_dst, buf->data,
-					buf->len);
-	}
-
 	if (get_opcode(buf, &opcode) < 0) {
 		LOG_WRN("Unable to decode OpCode");
 		return ACCESS_STATUS_WRONG_OPCODE;
@@ -1555,11 +1546,33 @@ int bt_mesh_model_recv(struct bt_mesh_msg_ctx *ctx, struct net_buf_simple *buf)
 		for (index = 0; index < dev_comp->elem_count; index++) {
 			const struct bt_mesh_elem *elem = &dev_comp->elem[index];
 
-			(void)element_model_recv(ctx, buf, elem, opcode);
+			int err_elem = element_model_recv(ctx, buf, elem, opcode);
+
+			err = err_elem == ACCESS_STATUS_SUCCESS ? err_elem : err;
 		}
 	}
 
+	return err;
+}
+
+int bt_mesh_model_recv(struct bt_mesh_msg_ctx *ctx, struct net_buf_simple *buf)
+{
+	int err;
+
+	LOG_DBG("app_idx 0x%04x src 0x%04x dst 0x%04x", ctx->app_idx, ctx->addr,
+		ctx->recv_dst);
+	LOG_DBG("len %u: %s", buf->len, bt_hex(buf->data, buf->len));
+
+	if (IS_ENABLED(CONFIG_BT_TESTING)) {
+		bt_test_mesh_model_recv(ctx->addr, ctx->recv_dst, buf->data,
+					buf->len);
+	}
+
+	err = bt_mesh_model_msg_recv(ctx, buf);
+
 	if (IS_ENABLED(CONFIG_BT_MESH_ACCESS_LAYER_MSG) && msg_cb) {
+		/* Overwritting error code as the message is processed by the callback. */
+		err = 0;
 		msg_cb(opcode, ctx, buf);
 	}
```

And the opagg models should use `bt_mesh_model_msg_recv` instead of `bt_mesh_model_recv`.

Reply:
The other option is to set some flag in both models, add a function to get the flag and call it when passing message to the access layer msg callback. Then encapsulated messages will also be passed to the callback.

Reply:
> > Didn't get what you mean. Please clarify.
> 
> When calling the access layer msg callback we need to check if an opagg model processes encapsulated messages now. If so, don't overwrite the error value.

There are no any properties in the implementation those notify access layer which functionality called its API. I guess this is not possible to do without API change. Or introduction the new API in oppcode aggregator model to let access layer to know about API usage. Sounds quite awkward to me, not better than analyzing `CONFIG_BT_MESH_ACCESS_LAYER_MSG` in transport layer.

If just return error code if oppcode aggregator models are present then we go back to replay attack scenario that I provided above.

> 
> > 1. Application doesn't provide any callback. This is possible. We do not mandate it.
> > 2. > Devices A1 have the disabled CONFIG_BT_MESH_ACCESS_LAYER_MSG at all.
> 
> In the first case messages are not processed, then it is OK to not store them. The second point is not a problem we are discussing in this thread. It is clear what to do in this case.
> 
> > Fair enough. We should mention it in KConfig option. Customers should know about potential problems.
> 
> It is not good to provide tons of documentation for issues that could be solved in the code.
> 

In general, I agree, but this is not an extraordinary issue. This might cause potentially to physical device damage.

> IMHO 1: If a user enabled the access layer msg callback option AND set the callback, the seqnums should be stored in RPL even if their processing would result in error defined by spec section 3.7.3.4. Because if we don't do this, then thus customers who rely only on the callback will see these messages in the callback over and over again.
> 

Yep, I agree with this and my implementation does it. The differences in solutions how to do it. Change error code on access or analyze `CONFIG_BT_MESH_ACCESS_LAYER_MSG ` option in transport.

If change error code on transport it causes a couple of situation and dependencies those should be solved separately (not clear if it is even possibly to solve).

My suggestion does it unconditionally of oppcode aggregator dependencies.

> IMHO 2: Introducing a dependency on the upper layer internal details (which is the access layer) in the lower layer (which is the transport layer) is not correct from the architecture perspective. The access layer shall return correct value which the transport layer should use to decide if the message was processed and it should be stored in RPL.

RPL lives apart from layers: 
```
3.9.8 Message replay protection
...
An implementation may perform the replay protection at any layer and in any order with respect to the message
 authentication steps (the network layer decryption and the transport layer decryption), in order to optimize the 
 message processing flow, the number of cryptographic operations or the memory usage. 
...
```
We trigger it in transport, but access also impacts on it. We have a lot such dependencies. Network depends on friendship, advertiser depends on friendship, proxy and network etc, etc

Reply:
Just to reach some decision without spending more time for discussion.

Please answer question. Do you really think that introduction of new access layer new function is better than just analyzing the single KConfig option on transport layer?

 If yes, I just add your suggestions despite I disagree with them.

Reply:
> There are no any properties in the implementation those notify access layer which functionality called its API. I guess this is not possible to do without API change. Or introduction the new API in oppcode aggregator model to let access layer to know about API usage. Sounds quite awkward to me, not better than analyzing `CONFIG_BT_MESH_ACCESS_LAYER_MSG` in transport layer.

Explained in the comment above. There is no need in API change. 
https://github.com/zephyrproject-rtos/zephyr/pull/70437#discussion_r1533598100
https://github.com/zephyrproject-rtos/zephyr/pull/70437#discussion_r1533612180

> If change error code on transport it causes a couple of situation and dependencies those should be solved separately (not clear if it is even possibly to solve).

Possible, explained above.

> My suggestion does it unconditionally of oppcode aggregator dependencies.
> RPL lives apart from layers:

RPL - yes, but I'm talking about the code interdependency, not about the spec.

> We trigger it in transport, but access also impacts on it. We have a lot such dependencies. Network depends on friendship, advertiser depends on friendship, proxy and network etc, etc

Depending on the lower layer is fine, not the opposite. It is a problem of the access layer **implementation** that it can't return correct error code. It is not the transport layer problem, therefore, should not be solved at the transport layer.

Reply:
> Just to reach some decision without spending more time for discussion.
> 
> Please answer question. Do you really think that introduction of new access layer function is better than just analyzing the single KConfig option on transport layer?
> 
> If yes, I just add your suggestions despite I disagree with them.

It looks simple now, it may create problems in future. As I said in the comment above, it is a problem of the access layer implementation that it can't return correct error code. It is not the transport layer problem, therefore, should not be solved at the transport layer.

Reply:
> > We trigger it in transport, but access also impacts on it. We have a lot such dependencies. Network depends on friendship, advertiser depends on friendship, proxy and network etc, etc
> 
> Depending on the lower layer is fine, not the opposite. It is a problem of the access layer **implementation** that it can't return correct error code. It is not the transport layer problem, therefore, should not be solved at the transport layer.

Which of the listed by me above dependencies are dependencies on the lower layers?

Friendship is transport layer functionality, but network and advertiser depend on it.
Relay is network functionality, but advertiser depends on it. 
and several more such dependencies.

My suggestion was: access layer callback is access layer functionality, but transport depends on it.

No dependencies on the lower layer in listed above.
Just would like to show your comment contradicts already existed implementation.

Reply:
> Which of the listed by me above dependencies are dependencies on the lower layers?
> 
> Friendship is transport layer functionality, but network and advertiser depend on it. Relay is network functionality, but advertiser depends on it. and several more such dependencies.

Not the topic of this thread. If you want, please bring this to team and we can discuss it there.

> My suggestion was: access layer callback is access layer functionality, but transport depends on it.
> 
> No dependencies on the lower layer in listed above. Just would like to show your comment contradicts already existed implementation.

Are there are other dependencies between transport and access layers? In code I mean.

Reply:
@PavelVPV, Seems it is not possible to get `opcode`  for access layer callback in `bt_mesh_model_recv`. Any ideas?

Reply:
https://github.com/PavelVPV/zephyr/commit/75f791148b19c2daa3e771173b300340b17868ba

Reply:
It doesn't seem a good solution if we have to create a global variable to keep state between steps. This mostly means either lack of normal state machine or bad design in general.

I guess the returned value from registered callback should improve design. Of course, it is changing of API, but it is better than crutches those we try to create.

Reply:
> It doesn't seem a good solution if we have to create a global variable to keep state between steps. This mostly means either lack of normal state machine or bad design in general.

We can use `bt_mesh.flags` for this I guess.

> I guess the returned value from registered callback should improve design. Of course, it is changing of API, but it is better than crutches those we try to create.

Please provide example.

Reply:
> @PavelVPV, Seems it is not possible to get `opcode` for access layer callback in `bt_mesh_model_recv`. Any ideas?

Or you can go with this approach and extract opcode for `msg_cb` like op_agg.c does before calling `bt_mesh_model_msg_recv`:
https://github.com/zephyrproject-rtos/zephyr/blob/b25eb36cb81917d529212c5a5c3d31ed56cd568d/subsys/bluetooth/mesh/op_agg.c#L73-L75

Reply:
Will this work: https://github.com/PavelVPV/zephyr/commit/b653141d580c625231ea9e0c8d2f562fdd7fc302 ?

Reply:
> > I guess the returned value from registered callback should improve design. Of course, it is changing of API, but it is better than crutches those we try to create.
> 
> Please provide example.

Please take a look. It is even not public API. Application can notify what mesh need to do over returned status. 
If it is success then it rpl is updated if it is not then depends on registered models. If all models rejected message then rpl is not updated either.

Also, it prevents scenario that I described when devices A2 actively communicating will wear out A1 persistent storage. 
Now A1 can notify mesh about that application didn't succeed with incoming message. rpl is not updated.

Reply:
If this is Ok, then it is possible to remove KConfig clarification since this is not the truth anymore.

Reply:
I see, nah, this is an API change. Then I suggest commit from this comment: https://github.com/zephyrproject-rtos/zephyr/pull/70437#discussion_r1535322271

Reply:
No, sorry, it is kind of not an API change since the registration function is inside `subsys`, but will affect customers who use this callback and we know there are some.

Though I like this more I guess this will still brake opagg logic. When an encapsulated message is passed to `bt_mesh_model_recv`, it will be passed to the callback too. The problem that you described above still exists. If a target model returned an error, the error will be overwritten by the callback here:
https://github.com/zephyrproject-rtos/zephyr/blob/764311b9712ce9e8ec0fe954a16c6e90a76796a5/subsys/bluetooth/mesh/access.c#L1569 and the opagg model will accept the message. The callback needs to know that this is an encapsulated message and should return anything other than 0 in this case.

Reply:
This behavior is out of spec. To be honest, I'm not sure what to do in this case. It seems like distributed model implementation. Part model is like a real model. Another part in located in app and accesses over app callback. Should oppcode aggregator handle such messages as a normal, or should it ignore them?

Reply:
I think that this depends on the device composition data. If the device composition data contains the certain model, then its messages should be processed when received through an aggregated message according to the opagg spec. If the device composition data does not contain the certain model and the device receives the encapsulated message it should reject it with WrongOpCode error code according to the opagg spec regardless of the callback as the callback is our "hack" to the access layer.